### PR TITLE
Add call to loop.Run to evaluate all setTimeout/setIntervals methods.

### PR DESCRIPTION
**Pull request by influx6**
## Issue
With the addition of the ottoext package which provides a event loop for usage with otto, we need to call `ottoext/loop.Loop.Run` method to ensure all pending tasks are runned within the loop, without this call, all tasks will be pending and unresolved.

See timer test and loop method:
1. https://github.com/deoxxa/ottoext/blob/master/loop/loop.go#L149
2. https://github.com/deoxxa/ottoext/blob/master/timers/timers_test.go#L17

### PR Adds
1. Removal of cloning with otto VM before use in executing `Call` request.
2. Addition of `loop.Run()` to ensure `setInterval/setTimeout/setImmediate` calls execute on loop.

### Future Problems:
After discussions with @tiabc and @adambabik, we currently are aware of the possibility of a gorountine leakage around calls to `loop.Run()` (see https://github.com/status-im/status-go/blob/2fba2b45eabc948f919b6944d14c643b76629c79/geth/jail/jail.go#L137). For now we will put this under observation and if needed, fork `ottoext` with needed changes to resolve this if we see any sign of mis-use or leakage.

Related to issue #196.

### CODE REVIEW

**Comment by rasom** [comment]
Let me know when I should test this in RN app.

**Comment by influx6** [comment]
Sure, will do.

When we have properly merged `bug/whisper-1.6.1` into develop, this will be tested. Sorry for the delay.

File: geth/jail/jail.go [hunk 1]
```diff
@@ -119,15 +119,23 @@ func (jail *Jail) Call(chatID string, path string, args string) string {
...
+	vm := cell.CellVM()
 	res, err := vm.Call("call", nil, path, args)
 
+	// Due to the new event loop provided by ottoext.
+	// We need to ensure that all possible calls to internal setIntervals/SetTimeouts/SetImmediate
+	// work by lunching the loop.Run() method.
+	// Needs to be done in a go-routine.
+	// TODO(influx6): Figure out if we should do a throw Exception for returned error
+	// from loop.Run() call.
+	go cell.CellLoop().Run()
```
Comment (tiabc) [inline geth/jail/jail.go]: It's going to be run every time jail.Call is invoked. Has it been supposed to be so?

Comment (influx6) [inline geth/jail/jail.go]: Yes, that is the idea, we have no means of knowing whether the call uses the loop, more so, we also are aware that, even if called if no task is pending the the event loop, it will not execute internal logic. So this should be safe.

Comment (tiabc) [inline geth/jail/jail.go]: But won't we get goroutines leakage with this? For the first glance, we will.

Comment (influx6) [inline geth/jail/jail.go]: Indeed, am worried about that myself.

Comment (tiabc) [inline geth/jail/jail.go]: So maybe we could run a single loop for every cell and execute commands on the associated vm? Will it work so?

Comment (influx6) [inline geth/jail/jail.go]: Hmm, each cell get's its own vm and the loop is required to be per vm, unless we are sharing a single vm to all cell which is not possible from current code, we can't have a single loop.

Comment (tiabc) [inline geth/jail/jail.go]: I mean, one loop per cell with a separate vm.
Because now we will have as many loops as Call is invoked for every cell.

Comment (influx6) [inline geth/jail/jail.go]: Oh not at all. Even cell has only one loop per vm, the loop exists already has part of the cell, but just called within the `Call` method.

See: https://github.com/status-im/status-go/blob/2fba2b45eabc948f919b6944d14c643b76629c79/geth/jail/jail.go#L52
See: https://github.com/status-im/status-go/blob/2fba2b45eabc948f919b6944d14c643b76629c79/geth/jail/jail_cell.go#L34-L52

_Unless I misunderstood your statement_

Comment (influx6) [inline geth/jail/jail.go]: The loop itself implements the jail `setTimeout/setInterval` methods attached to the vm, hence it's the one which such tasks are appended to, `loop.Run`, simple asks the loop to execute those tasks. 

It's similar to `EvalAndRun` method in https://github.com/deoxxa/ottoext/blob/master/loop/loop.go#L111.

Comment (influx6) [inline geth/jail/jail.go]: We could use `EvalAndRun` but be aware that will block till all tasks are resolved, and a setInterval, we continously add tasks into the queue.

Comment (influx6) [inline geth/jail/jail.go]: Also, since the vm cloning has been removed, there are no new vm or loops during the `Call` method.

Comment (influx6) [inline geth/jail/jail.go]: See https://github.com/status-im/status-go/blob/2fba2b45eabc948f919b6944d14c643b76629c79/geth/jail/jail.go#L120-L125

Comment (adambabik) [inline geth/jail/jail.go]: I just glimpsed at ottoext code and wonder if you run the loop in a goroutine and then each new code via `Loop.Eval()`, then new timers won't be added to the existing loop?

Comment (influx6) [inline geth/jail/jail.go]: Yes, we can run `loop.Run` in one go-routine, then allow running of tasks when setIntervals calls, but the Run will default at initial run if no tasks are pending.

Comment (influx6) [inline geth/jail/jail.go]: We could however, fork and add a select to always await new tasks, but also ensure we equally take care to kill goroutine when cell is removed, which currently i don't think we are catering to.

Comment (adambabik) [inline geth/jail/jail.go]: Ah right, because the loop terminates if there are no pending tasks. So, I guess we need to keep track if the loop is running. If it's not, eval with `Loop.EvalAndRun()`, otherwise just `Loop.Eval()`. Or am I missing something?

Comment (tiabc) [inline geth/jail/jail.go]: After discussion in slack:

We could fork it and it may be a good decision. For now let's decide to leave it as-is. We may face problems with it in case many setInterval's run but it's the choice and responsibility of the user to run them.
If we face any problems with it, we'll either fork as you proposed, or make some time limits for execution, for example.
Please, add a comment warning that we can have a goroutine leakage in this place in case of improper use.

File: geth/jail/jail.go [hunk 2]
```diff
@@ -46,17 +46,39 @@ func (jail *Jail) BaseJS(js string) {
...
 
 	newJail, err := newJailCell(id, vm, loop.New(vm))
 	if err != nil {
-		//TODO(alex): Should we really panic here, his there
-		// a better way. Think on it.
-		panic(err)
+		return nil, err
+	}
+
+	jail.RLock()
```
Comment (adambabik) [inline geth/jail/jail.go]: `jail.cells[id] = newJail` operation requires write lock. So you should do `jail.Lock()`. Also, putting it in a block is not necessary.

File: geth/jail/jail.go [hunk 3]
```diff
@@ -46,17 +46,39 @@ func (jail *Jail) BaseJS(js string) {
...
 	}
+	jail.RUnlock()
 
-	return newJail
+	return newJail, nil
+}
+
+// GetCell returns instance of jailed runtime
+func (jail *Jail) GetCell(chatID string) (common.JailCell, error) {
+
```
Comment (adambabik) [inline geth/jail/jail.go]: Blank line not needed.

File: geth/jail/jail.go [hunk 4]
```diff
@@ -46,17 +46,39 @@ func (jail *Jail) BaseJS(js string) {
 }
 
 // NewJailCell initializes and returns jail cell
-func (jail *Jail) NewJailCell(id string) common.JailCell {
+func (jail *Jail) NewJailCell(id string) (common.JailCell, error) {
```
Comment (adambabik) [inline geth/jail/jail.go]: You should return a specific type here, not the interface.

File: geth/jail/jail.go [hunk 5]
```diff
@@ -67,22 +89,28 @@ func (jail *Jail) Parse(chatID string, js string) string {
...
 	}
 
-	jail.Lock()
-	defer jail.Unlock()
+	cell, err := jail.NewJailCell(chatID)
+	if err != nil {
+		return makeError(err.Error())
+	}
+
+	jcell, _ := cell.(*JailCell)
```
Comment (adambabik) [inline geth/jail/jail.go]: If `jail.NewJailCell()` returns a specific type, this line won't be needed.

File: geth/jail/jail.go [hunk 6]
```diff
@@ -67,22 +89,28 @@ func (jail *Jail) Parse(chatID string, js string) string {
...
 
-	jail.cells[chatID] = jail.NewJailCell(chatID)
-	vm := jail.cells[chatID].CellVM()
+	jcell.Lock()
+	defer jcell.Unlock()
 
 	initJjs := jail.baseJSCode + ";"
-	if _, err = vm.Run(initJjs); err != nil {
+	if _, err = jcell.vm.Run(initJjs); err != nil {
```
Comment (adambabik) [inline geth/jail/jail.go]: Maybe we can fix `initJjs` variable. It looks like a typo.

File: geth/jail/jail_cell.go [hunk 7]
```diff
@@ -51,19 +49,17 @@ func newJailCell(id string, vm *otto.Otto, lo *loop.Loop) (*JailCell, error) {
...
-	return newJailCell(cell.id, vmCopy, loop.New(vmCopy))
-}
-
 // Fetch attempts to call the underline Fetch API added through the
 // ottoext package.
 func (cell *JailCell) Fetch(url string, callback func(otto.Value)) (otto.Value, error) {
-	if err := cell.vm.Set("__captureFetch", callback); err != nil {
-		return otto.UndefinedValue(), err
+	cell.Lock()
+	{
```
Comment (adambabik) [inline geth/jail/jail_cell.go]: What is this block for?

File: geth/jail/jail_cell.go [hunk 8]
```diff
@@ -90,21 +89,25 @@ func (cell *JailCell) Exec(val string) (otto.Value, error) {
...
+
 	return cell.vm.Run(val)
 }
 
-// CellLoop returns the ottoext.Loop instance which provides underline timeout/setInternval
+// Loop returns the ottoext.Loop instance which provides underline timeout/setInternval
 // event runtime for the Jail vm.
-func (cell *JailCell) CellLoop() *loop.Loop {
+func (cell *JailCell) Loop() *loop.Loop {
+	cell.Lock()
```
Comment (adambabik) [inline geth/jail/jail_cell.go]: You can use `cell.RLock()` here.

File: geth/jail/jail_cell.go [hunk 9]
```diff
@@ -90,21 +89,25 @@ func (cell *JailCell) Exec(val string) (otto.Value, error) {
...
 	return cell.lo
 }
 
-// Executor returns a structure which implements the common.JailExecutor.
-func (cell *JailCell) Executor() common.JailExecutor {
-	return cell
-}
+// VM returns the associated otto.Vm connect to the giving cell.
+func (cell *JailCell) VM() *otto.Otto {
+	cell.Lock()
```
Comment (adambabik) [inline geth/jail/jail_cell.go]: Here `.RLock()` should be fine as well.

File: geth/jail/jail_test.go [hunk 10]
```diff
@@ -196,6 +197,73 @@ func (s *JailTestSuite) TestJailFetch() {
...
+		}
+	`)
+	require.NoError(err)
+
+	// NOTE: Fails to work because it's running intirely
+	_, err = cellVM.Call("callRunner", nil, "softball")
+	require.NoError(err)
+
+	go func() {
+		require.NoError(cellLoop.Run())
```
Comment (adambabik) [inline geth/jail/jail_test.go]: You could also write `callRunner()` to finish after a few iterations and test if the loop actually finishes as well.

File: geth/common/types.go [hunk 11]
```diff
@@ -173,28 +173,11 @@ type TxQueueManager interface {
...
-	Fetch(string, func(otto.Value)) (otto.Value, error)
-}
-
 // JailCell represents single jail cell, which is basically a JavaScript VM.
 type JailCell interface {
-	CellVM() *otto.Otto
-	CellLoop() *loop.Loop
-	Executor() JailExecutor
-	Copy() (JailCell, error)
+	VM() *otto.Otto
```
Comment (adambabik) [inline geth/common/types.go]: Nice refactoring 👍

File: geth/jail/jail.go (hunk 4 above)
Comment (influx6) [inline geth/jail/jail.go]: If we do so, we will not be able to meet the `JailManager` interface.  See https://github.com/status-im/status-go/blob/77d01c424e5dab818ce07cff38bdd9779ecdc2c8/geth/common/types.go#L194

File: geth/jail/jail_cell.go (hunk 8 above)
Comment (influx6) [inline geth/jail/jail_cell.go]: The synchronization lock being used here is a `sync.Mutex` except you wish it to be changed to a `sync.RWMutex`.

File: geth/jail/jail_cell.go (hunk 9 above)
Comment (influx6) [inline geth/jail/jail_cell.go]: The synchronization lock being used here is a `sync.Mutex` except you wish it to be changed to a `sync.RWMutex`.

File: geth/jail/jail_test.go (hunk 10 above)
Comment (influx6) [inline geth/jail/jail_test.go]: Hi @adambabik, am not sure I understand the requirement here. Since the `cellLoop.Run` blocks, it was placed into the goroutine. Running `cellLoop.Run` multiple times would not be suited has each call will block has far as items are still awaiting processing within the loop. 

Hence calls to `cellLoop.Run()` must be called either:

1. Immediately after a call to `Otto.Call`, `Otto.Run` to execute any pending tasks that made use of `setTimeout/setInterval/setImmediate`.

2. After stacking multiple calls to  `Otto.Call/Otto.Run` then a call to `cellLoop.Run()` to batch handle all currently stacked requests.

If you mean adding a test after validating that the items where fully received to validate no item is currently pending in the `cellLoop`, then I can do that.

File: geth/jail/jail.go (hunk 4 above)
Comment (adambabik) [inline geth/jail/jail.go]: Oh I see. I haven't noticed that. Let's leave it as it's then.

File: geth/jail/jail_test.go (hunk 10 above)
Comment (adambabik) [inline geth/jail/jail_test.go]: > If you mean adding a test after validating that the items where fully received to validate no item is currently pending in the cellLoop, then I can do that.

This is exactly what I mean. On the other hand, we have a `TODO` comment that we need to handle the loop and any possible errors. So, maybe we can add additional tests when we start fixing it instead of spending time now. 

TLDR; let's leave it as it is for the time being.

Comment (influx6) [inline geth/jail/jail_test.go]: Sounds good.

**Comment by rasom** [comment]
I also get this error when trying to compile for android:
```
../../geth/api/backend.go:36: cannot use jail.New(nodeManager) (type *jail.Jail) as type "github.com/status-im/status-go/geth/common".JailManager in field value:
	*jail.Jail does not implement "github.com/status-im/status-go/geth/common".JailManager (wrong type for NewJailCell method)
		have NewJailCell(string) ("github.com/status-im/status-go/geth/common".JailCell, error)
		want NewJailCell(string) "github.com/status-im/status-go/geth/common".JailCell
2017/08/01 15:27:07 Failed to cross compile package: exit status 2.
make: *** [statusgo-android] Error 1
```

**Comment by influx6** [comment]
Looking into this now.

**Comment by influx6** [comment]
@rasom Please `pull --rebase` and test again. Build works successfully.

**Comment by rasom** [comment]
Library can be compiled now, thanks @influx6.

File: geth/common/types.go (hunk 11 above)
Comment (tiabc) [inline geth/common/types.go]: Do we really need this interface? If we remove it, we lose nothing because it's only returned in tests and used nowhere as an argument. We'll also be able to remove `Jail.GetCell` and rename `Jail.getCell` into `Jail.GetCell`.

As I can see, its only purpose is to encapsulate `vm` and `loop` fields and:
1. It is not a recommended practice in Go.
2. It adds unnecessary complexity.
3. `VM` and `Loop` methods are only called in unit tests which can easily be replaced with "real" fields.
4. It will allow you to get rid of unneeded type assertions.

File: geth/jail/jail.go [hunk 12]
```diff
@@ -109,60 +139,49 @@ func (jail *Jail) Parse(chatID string, js string) string {
...
+	// Needs to be done in a go-routine.
+	// TODO(influx6): Figure out if we should do a throw Exception for returned error
+	// from loop.Run() call.
+	go loop.Run()
+
 	return makeResult(res.String(), err)
 }
 
 // JailCellVM returns instance of Otto VM (which is persisted w/i jail cell) by chatID
+// TODO(influx6): Decide if this make sense now that we are guarding JailCell vm with mutex.
```
Comment (tiabc) [inline geth/jail/jail.go]: Please, remove this method and fix tests. It's only used there.
Moreover, we shouldn't expose `vm` and `loop` to other parties because we need to guarantee that all calls are made in sequence so you'll have to additionally check that only JailCell methods are used, not its `vm` or `loop`.

File: geth/jail/jail_cell.go [hunk 13]
```diff
@@ -51,20 +49,18 @@ func newJailCell(id string, vm *otto.Otto, lo *loop.Loop) (*JailCell, error) {
...
-	vmCopy := cell.vm.Copy()
-	return newJailCell(cell.id, vmCopy, loop.New(vmCopy))
-}
-
 // Fetch attempts to call the underline Fetch API added through the
 // ottoext package.
 func (cell *JailCell) Fetch(url string, callback func(otto.Value)) (otto.Value, error) {
+	cell.Lock()
 	if err := cell.vm.Set("__captureFetch", callback); err != nil {
+		defer cell.Unlock()
```
Comment (tiabc) [inline geth/jail/jail_cell.go]: That looks strange. Move it right after `cell.Lock()`?

File: geth/common/types.go (hunk 11 above)
Comment (influx6) [inline geth/common/types.go]: To remove `JailCell` interface, we will need to sure that `JailManager.NewJailCell` is not executed anywhere but from internal logic, else remove `NewJailCell` from `JailManager` interface.

Comment (influx6) [inline geth/common/types.go]: More so, I proprose we remove the `JailCellVM` from the `JailManager` has that is breaking the purpose of having `jail.JailCell` to manage concurrent access to the internal `otto.Otto` vm.

Comment (influx6) [inline geth/common/types.go]: @tiabc But cases like this https://github.com/status-im/status-go/blob/77d01c424e5dab818ce07cff38bdd9779ecdc2c8/geth/api/backend_jail_test.go#L615-L622, would make that impossible, without first doing a type asset for the `JailManager` interface.

File: geth/jail/jail.go (hunk 12 above)
Comment (influx6) [inline geth/jail/jail.go]: We will need to expose atleast a `GetJailCell` method for the needs of external packages that have access to the `common.JailManager` like in test https://github.com/status-im/status-go/blob/77d01c424e5dab818ce07cff38bdd9779ecdc2c8/geth/api/backend_jail_test.go#L615-L622. But since this is test, we can easily type assert here, and not need to expose such a method.

File: geth/jail/jail_cell.go (hunk 13 above)
Comment (influx6) [inline geth/jail/jail_cell.go]: Seen it, the defer is supposed to be removed here. Will fix that.

**Comment by influx6** [comment]
Should we rebase the large commit tree and squash them?

File: geth/common/types.go (hunk 11 above)
Comment (tiabc) [inline geth/common/types.go]: Can you, please, check this?

File: geth/jail/jail.go (hunk 12 above)
Comment (tiabc) [inline geth/jail/jail.go]: Yes, absolutely.

File: geth/jail/jail_cell.go (hunk 13 above)
Comment (tiabc) [inline geth/jail/jail_cell.go]: I would rather leave it but moved right after `cell.Lock()` just because you won't forget to do `Unlock` in case you create another branch or something.

**Comment by tiabc** [comment]
@influx6, I think we'll squash them before merging into the original branch.

File: geth/common/types.go (hunk 11 above)
Comment (influx6) [inline geth/common/types.go]: Resolved in latest commit

**Comment by tiabc** [comment]
Actually, no. I can see that there're commits in this pull request which are already in the original branch and if you look at git graph, you'll see that you just did bad rebase. Please, fix.

File: geth/jail/jail.go [hunk 14]
```diff
@@ -67,28 +91,34 @@ func (jail *Jail) Parse(chatID string, js string) string {
 		return makeError(ErrInvalidJail.Error())
 	}
 
-	jail.Lock()
-	defer jail.Unlock()
+	//TODO(influx6): Find out if calls to Parse must always start with new Jail,
+	// is it possible to get a ChatID that was previous registered?
+	if _, err := jail.NewJailCell(chatID); err != nil {
+		return makeError(err.Error())
+	}
```
Comment (tiabc) [inline geth/jail/jail.go]: Good question. It's worth investigating right now.

File: geth/jail/jail.go [hunk 15]
```diff
@@ -67,28 +91,34 @@ func (jail *Jail) Parse(chatID string, js string) string {
...
-	jail.cells[chatID] = jail.NewJailCell(chatID)
-	vm := jail.cells[chatID].CellVM()
+	jcell, _ := jail.GetCell(chatID)
+
+	jcell.Lock()
+	defer jcell.Unlock()
 
 	// init jeth and its handlers
-	if err = vm.Set("jeth", struct{}{}); err != nil {
+	if err = jcell.vm.Set("jeth", struct{}{}); err != nil {
```
Comment (tiabc) [inline geth/jail/jail.go]: Candidate for `jcell.Set`?

File: geth/jail/jail_cell.go [hunk 16]
```diff
@@ -43,29 +36,26 @@ func newJailCell(id string, vm *otto.Otto, lo *loop.Loop) (*JailCell, error) {
...
-func (cell *JailCell) Copy() (common.JailCell, error) {
-	vmCopy := cell.vm.Copy()
-	return newJailCell(cell.id, vmCopy, loop.New(vmCopy))
-}
-
 // Fetch attempts to call the underline Fetch API added through the
 // ottoext package.
-func (cell *JailCell) Fetch(url string, callback func(otto.Value)) (otto.Value, error) {
+func (cell *JCell) Fetch(url string, callback func(otto.Value)) (otto.Value, error) {
+	cell.Lock()
```
Comment (tiabc) [inline geth/jail/jail_cell.go]: `defer cell.Unlock()` looks more reliable, right?

File: geth/jail/jail_cell.go [hunk 17]
```diff
@@ -1,38 +1,31 @@
...
+	// JCellRequestTimeout seconds before jailed request times out.
+	JCellRequestTimeout = 60
 )
 
-// JailCell represents single jail cell, which is basically a JavaScript VM.
-type JailCell struct {
-	// FIXME(tiabc): It's never called. Double-check and delete.
-	//sync.Mutex
+// JCell represents single jail cell, which is basically a JavaScript VM.
+type JCell struct {
```
Comment (tiabc) [inline geth/jail/jail_cell.go]: Please, refrain from refactoring in regular tasks: https://hackmd.io/s/rkgqzRWIb

You can add a `TODO` if you want to rename it under some refactoring later.

File: geth/jail/jail.go [hunk 18]
```diff
@@ -46,17 +46,41 @@ func (jail *Jail) BaseJS(js string) {
...
+	jail.RLock()
+	defer jail.RUnlock()
+
+	cell, ok := jail.cells[chatID]
+	if !ok {
+		return nil, fmt.Errorf("cell[%s] doesn't exist", chatID)
+	}
+
+	return cell, nil
 }
```
Comment (tiabc) [inline geth/jail/jail.go]: Are you sure we can't get rid of one of `GetJailCell` and `GetCell`?

Comment (influx6) [inline geth/jail/jail.go]: Unfortunately no, there are cases where outside packages need access to the common.JailCell either in tests or in code. Removing this would break that and leave rather ugly type assertion.

Allow GetCell on the struct value solves the issue inside the jail package as regards receing proper solid type.

Unless you are alright with type assertions in different packages when Jail is being used, this is currently the best fit.

File: geth/jail/jail_cell.go (hunk 17 above)
Comment (influx6) [inline geth/jail/jail_cell.go]: Alright. Understood!

File: geth/jail/jail_cell.go (hunk 16 above)
Comment (influx6) [inline geth/jail/jail_cell.go]: Not possible, as the `jcell.Run` acquires the lock as well. If we want `defer` then i will swap out proper direct calls in here as well.

**Comment by influx6** [comment]
Yes there are cases where the underline `JCell` internal lock and vm are access directly. The purpose was to leave tge areas that need those calls unchanged but guarded directly with the mutex.

That can change but the speific case of the `Jail.Call` require a more method on `JCell` which will be goroutined on the call for its loop to run tasks.

If you wish, i will refactor out the direct access and let things in this few last places as well.

File: geth/jail/jail.go (hunk 18 above)
Comment (tiabc) [inline geth/jail/jail.go]: No, I'm not alright with type assertions. Nor with methods looking like duplicated. Let's leave it as-is for now because it's out of scope of this issue but I'll include it in a refactoring issue.

File: geth/jail/jail_cell.go (hunk 16 above)
Comment (tiabc) [inline geth/jail/jail_cell.go]: I suggest we add `defer` and call `cell.vm.Run` at the end as this combination looks more straightforward for understanding and less error-prone.

**Comment by tiabc** [comment]
@influx6, let's leave refactoring until better days.

File: geth/jail/jail_cell.go [hunk 19]
```diff
@@ -65,6 +63,12 @@ func (cell *JCell) Fetch(url string, callback func(otto.Value)) (otto.Value, err
 			});
 		});
 	`)
+
+	if err != nil {
+		return val, err
+	}
+
+	return val, cell.lo.Run()
```
Comment (influx6) [inline geth/jail/jail_cell.go]: This must be called here to ensure fetch executes immediately in the event loop.

File: geth/jail/jail.go [hunk 20]
```diff
@@ -115,60 +144,26 @@ func (jail *Jail) Parse(chatID string, js string) string {
...
-}
+	res, err := jcell.Call("call", nil, path, args)
 
-// GetCell returns instance of jailed runtime
-func (jail *Jail) GetCell(chatID string) (common.JailCell, error) {
-	if jail == nil {
-		return nil, ErrInvalidJail
-	}
+	jcell.Lock()
+	defer jcell.Unlock()
```
Comment (tiabc) [inline geth/jail/jail.go]: Hm. Why are we acquiring lock here, by the way? `jcell.Call` above already does locking inside as well as `lo.Run` below.

File: geth/jail/jail_cell_test.go [hunk 21]
```diff
@@ -1,3 +1,135 @@
...
+	`)
+
+	require.NoError(err)
+	require.NotNil(res)
+}
+
+// TODO(influx6): JCell.Fetch is not yet being used or needed by
+// other API, keep test out for the main time, if method is later required
+// add to common.JailCell interface then uncomment test.
+// func (s *JailTestSuite) TestJailFetch() {
```
Comment (tiabc) [inline geth/jail/jail_cell_test.go]: If we implemented fetch, we must write a test for it. Please, uncomment. There has already been a request from one person from the community regarding `fetch` so it may immediately be used after the release.

File: geth/jail/jail.go (hunk 20 above)
Comment (influx6) [inline geth/jail/jail.go]: Due to the need of access `lo`. But will remove this then.

File: geth/jail/jail_cell.go [hunk 22]
```diff
@@ -75,11 +64,39 @@ func (cell *JailCell) Fetch(url string, callback func(otto.Value)) (otto.Value,
 			});
 		});
 	`)
+
+	if err != nil {
+		cell.Unlock()
+		return val, err
+	}
+
+	cell.Unlock()
```
Comment (tiabc) [inline geth/jail/jail_cell.go]: Let's move all that internal logic but the last return to a private function. So that `defer cell.Unlock()` is reliably called inside it.

Comment (influx6) [inline geth/jail/jail_cell.go]: Alright.
